pyk/src/tests/integration/proof/test_custom_step.py

```diff
@@ -0,0 +1,232 @@
+from __future__ import annotations
+
+from pathlib import Path
+from typing import TYPE_CHECKING
+
+import pytest
+
+from pyk.cterm import CTerm
+from pyk.kast.inner import KApply, KLabel, KSequence, KSort, KVariable
+from pyk.kast.manip import set_cell
+from pyk.kcfg import KCFGExplore
+from pyk.kcfg.kcfg import Step
+from pyk.kcfg.semantics import KCFGSemantics
+from pyk.kcfg.show import KCFGShow
+from pyk.proof import APRProof, APRProver, ProofStatus
+from pyk.proof.show import APRProofNodePrinter
+from pyk.testing import CTermSymbolicTest, KProveTest
+from pyk.utils import single
+
+from ..utils import K_FILES
+
+if TYPE_CHECKING:
+    from collections.abc import Callable, Iterable
+    from typing import Union
+
+    from pyk.cterm import CTermSymbolic
+    from pyk.kast.outer import KClaim
+    from pyk.kcfg.kcfg import KCFGExtendResult
+    from pyk.ktool.kprove import KProve
+    from pyk.utils import BugReport
+
+    STATE = Union[tuple[str, str], tuple[str, str, str]]
+
+
+CUSTOM_STEP_TEST_DATA: Iterable[tuple[str, Path, str, str, int | None, int | None, Iterable[str], ProofStatus]] = (
+    (
+        'a-to-d',
+        K_FILES / 'custom-step-spec.k',
+        'CUSTOM-STEP-SPEC',
+        'a-to-d',
+        3,
+        4,
+        ['CUSTOM-STEP.c.d'],
+        ProofStatus.PASSED,
+    ),
+)
+
+
+class CustomStepSemanticsWithoutStep(KCFGSemantics):
+    def is_terminal(self, c: CTerm) -> bool:
+        k_cell = c.cell('K_CELL')
+        if (
+            type(k_cell) is KSequence
+            and type(k_cell[0]) is KApply
+            and k_cell[0].label.name == 'd_CUSTOM-STEP-SYNTAX_Step'
+        ):
+            return True
+        return False
+
+    def abstract_node(self, c: CTerm) -> CTerm:
+        return c
+
+    def same_loop(self, c1: CTerm, c2: CTerm) -> bool:
+        return False
+
+    def custom_step(self, c: CTerm) -> KCFGExtendResult | None:
+        return None
+
+
+class CustomStepSemanticsWithStep(CustomStepSemanticsWithoutStep):
+    def custom_step(self, c: CTerm) -> KCFGExtendResult | None:
+        k_cell = c.cell('K_CELL')
+        if (
+            type(k_cell) is KSequence
+            and type(k_cell[0]) is KApply
+            and k_cell[0].label.name == 'c_CUSTOM-STEP-SYNTAX_Step'
+        ):
+            new_cterm = CTerm.from_kast(set_cell(c.kast, 'K_CELL', KSequence(KApply('d_CUSTOM-STEP-SYNTAX_Step'))))
+            return Step(new_cterm, 1, (), ['CUSTOM-STEP.c.d'], cut=True)
+        return None
+
+
+def cterm_template(label: str) -> CTerm:
+    return CTerm(
+        KApply(
+            KLabel('<generatedTop>'),
+            [
+                KApply(
+                    KLabel('<k>'),
+                    [
+                        KSequence(KApply(KLabel(label))),
+                    ],
+                ),
+                KApply(KLabel('<generatedCounter>'), [KVariable('GENERATEDCOUNTER_CELL', KSort(name='Int'))]),
+            ],
+        ),
+    )
```

Comment:
You can make this a bit more succinct:

```suggestion
    return CTerm(
        KApply(
            '<generatedTop>',
            KApply('<k>', KSequence(KApply(label))),
            KApply('<generatedCounter>', KVariable('GENERATEDCOUNTER_CELL', 'Int')),
        ),
    )
```